core-primitives/top-pool-author/src/api.rs

```diff
@@ -19,92 +19,74 @@
 use sp_runtime::{
 	generic::BlockId,
-	traits::{Block as BlockT, Hash as HashT, Header as HeaderT},
-	transaction_validity::{
-		TransactionValidity, TransactionValidityError, UnknownTransaction, ValidTransaction,
-	},
+	traits::{BlakeTwo256, Block as BlockT, Hash as HashT},
+	transaction_validity::TransactionValidity,
 };
-use std::{boxed::Box, marker::PhantomData, pin::Pin, vec, vec::Vec};
+use std::{boxed::Box, marker::PhantomData, pin::Pin};
 
 /// Future that resolves to account nonce.
 pub type Result<T> = core::result::Result<T, ()>;
 
 /// The operation pool logic for full client.
-pub struct SidechainApi<Block> {
-	_marker: PhantomData<Block>,
+pub struct SidechainApi<Block, TCS> {
+	_marker: PhantomData<(Block, TCS)>,
 }
 
-impl<Block> SidechainApi<Block> {
+impl<Block, TCS> SidechainApi<Block, TCS>
+where
+	TCS: PartialEq + TrustedCallVerification + Debug,
+{
 	/// Create new operation pool logic.
 	pub fn new() -> Self {
 		SidechainApi { _marker: Default::default() }
 	}
-
-	fn validate_trusted_call(trusted_call_signed: TrustedCallSigned) -> ValidTransaction {
-		let from = trusted_call_signed.call.sender_account();
-		let requires = vec![];
-		let provides = vec![(from, trusted_call_signed.nonce).encode()];
-
-		ValidTransaction { priority: 1 << 20, requires, provides, longevity: 64, propagate: true }
-	}
 }
 
-impl<Block> Default for SidechainApi<Block> {
+impl<Block, TCS> Default for SidechainApi<Block, TCS>
+where
+	TCS: PartialEq + TrustedCallVerification + Debug + Sync + Send,
+{
 	fn default() -> Self {
 		Self::new()
 	}
 }
 
-impl<Block> ChainApi for SidechainApi<Block>
+impl<Block, TCS> ChainApi for SidechainApi<Block, TCS>
 where
 	Block: BlockT,
+	TCS: PartialEq + TrustedCallVerification + Sync + Send + Debug,
 {
 	type Block = Block;
 	type Error = error::Error;
 	type ValidationFuture =
 		Pin<Box<dyn Future<Output = error::Result<TransactionValidity>> + Send>>;
-	type BodyFuture = Ready<error::Result<Option<Vec<StfTrustedOperation>>>>;
+	type BodyFuture = Ready<error::Result<Option<bool>>>;
 
-	fn validate_transaction(
+	fn validate_transaction<TOP: PoolTransactionValidation>(
 		&self,
 		_source: TrustedOperationSource,
-		uxt: StfTrustedOperation,
+		uxt: TOP,
 		_shard: ShardIdentifier,
 	) -> Self::ValidationFuture {
-		let operation = match uxt {
-			StfTrustedOperation::direct_call(signed_call) =>
-				Self::validate_trusted_call(signed_call),
-			StfTrustedOperation::indirect_call(signed_call) =>
-				Self::validate_trusted_call(signed_call),
-			StfTrustedOperation::get(getter) => match getter {
-				Getter::public(_) =>
-					return Box::pin(ready(Ok(Err(TransactionValidityError::Unknown(
-						UnknownTransaction::CannotLookup,
-					))))),
-				Getter::trusted(trusted_getter) => ValidTransaction {
-					priority: 1 << 20,
-					requires: vec![],
-					provides: vec![trusted_getter.signature.encode()],
-					longevity: 64,
-					propagate: true,
-				},
-			},
-		};
-		Box::pin(ready(Ok(Ok(operation))))
```

Comment:
moving that into the generic types causes a slightly different behaviour and some unit tests had to be adjusted. over all, I think it is not dramatic, but as the verification doesn't happen at exactly the same point, some interna needed to be tested differently (i.e. nonce and signature checks, treating failed tops in pool....